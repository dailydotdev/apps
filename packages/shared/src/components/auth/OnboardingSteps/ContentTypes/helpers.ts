import { AdvancedSettings } from '../../../../graphql/feedSettings';
import {
  getContentCurationList,
  getContentSourceList,
  getVideoSetting,
} from '../../../filters/helpers';
import { Source } from '../../../../graphql/sources';

interface GetContentTypeNotEmptyProps {
  advancedSettings: AdvancedSettings[];
  selectedSettings: Record<string, boolean>;
  checkSourceBlocked: (source: Source) => boolean;
}

export const getContentTypeNotEmpty = ({
  advancedSettings,
  selectedSettings,
  checkSourceBlocked,
}: GetContentTypeNotEmptyProps): boolean => {
  const contentSourceList = getContentSourceList(advancedSettings);
  const contentCurationList = getContentCurationList(advancedSettings);

  const videoSetting = getVideoSetting(advancedSettings);

  const advancedSettingsSelected = (settings: AdvancedSettings[]) =>
    settings
      .map(({ id, defaultEnabledState }) => {
        return selectedSettings[id] ?? defaultEnabledState;
      })
      .some((setting) => setting === true);

  const advancedSettingsCurationListSelected =
    advancedSettingsSelected(contentCurationList);

  const advancedSettingsVideoSelected =
    !!videoSetting && advancedSettingsSelected([videoSetting]);

  const sourceListSelected = contentSourceList
    .map(({ options }) => options.source)
    .some((source) => !checkSourceBlocked(source));

  return (
    advancedSettingsCurationListSelected ||
    advancedSettingsVideoSelected ||
    sourceListSelected
  );
};
