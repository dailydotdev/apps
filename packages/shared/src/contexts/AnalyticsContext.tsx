import React, { ReactElement, ReactNode, createContext, useRef } from 'react';
import useAnalyticsQueue, {
  AnalyticsEvent,
} from '../hooks/analytics/useAnalyticsQueue';
import useAnalyticsSharedProps from '../hooks/analytics/useAnalyticsSharedProps';
import useAnalyticsContextData, {
  AnalyticsContextData,
} from '../hooks/analytics/useAnalyticsContextData';
import useBackfillPendingEvents from '../hooks/analytics/useBackfillPendingEvents';
import useTrackLifecycleEvents from '../hooks/analytics/useTrackLifecycleEvents';

const AnalyticsContext = createContext<AnalyticsContextData>({
  trackEvent: () => {},
  trackEventStart: () => {},
  trackEventEnd: () => {},
});
export default AnalyticsContext;

export type AnalyticsContextProviderProps = {
  app: string;
  getPage: () => string;
  version?: string;
  fetchMethod?: typeof fetch;
  backgroundMethod?: (msg: unknown) => Promise<unknown>;
  deviceId?: string;
  children?: ReactNode;
};

export const AnalyticsContextProvider = ({
  app,
  version,
  children,
  fetchMethod = fetch,
  backgroundMethod,
  deviceId,
  getPage,
}: AnalyticsContextProviderProps): ReactElement => {
  const { pushToQueue, setEnabled, queueRef, sendBeacon } = useAnalyticsQueue({
    fetchMethod,
    backgroundMethod,
  });
  const [sharedPropsRef, sharedPropsSet] = useAnalyticsSharedProps(
    app,
    version,
    deviceId,
  );
  const durationEventsQueue = useRef<Map<string, AnalyticsEvent>>(new Map());
  const contextData = useAnalyticsContextData(
    pushToQueue,
    sharedPropsRef,
    getPage,
    durationEventsQueue,
  );
  useBackfillPendingEvents(
    sharedPropsRef,
    sharedPropsSet,
    queueRef,
    durationEventsQueue,
    setEnabled,
  );
  useTrackLifecycleEvents(
    setEnabled,
    contextData,
    durationEventsQueue,
    sendBeacon,
  );

  return (
    <AnalyticsContext.Provider value={contextData}>
      {children}
    </AnalyticsContext.Provider>
  );
};
