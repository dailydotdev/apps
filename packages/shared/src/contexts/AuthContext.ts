import React from 'react';
import { LoggedUser } from '../lib/user';
import { LoginModalMode } from '../types/LoginModalMode';

export interface AuthContextData {
  user: LoggedUser;
  shouldShowLogin: boolean;
  showLogin: (mode?: LoginModalMode) => void;
  logout: () => Promise<void>;
  updateUser: (user: LoggedUser) => Promise<void>;
  loadingUser?: boolean;
  tokenRefreshed: boolean;
  loadedUserFromCache?: boolean;
  getRedirectUri: () => string;
}

const AuthContext = React.createContext<AuthContextData>(null);
export default AuthContext;
