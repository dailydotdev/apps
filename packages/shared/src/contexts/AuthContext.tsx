import React, { ReactElement, ReactNode, useMemo, useState } from 'react';
import {
  AnonymousUser,
  LoggedUser,
  logout as dispatchLogout,
  deleteAccount,
} from '../lib/user';
import { Visit } from '../lib/boot';

export type LoginState = { trigger: string };

export interface AuthContextData {
  user?: LoggedUser;
  trackingId?: string;
  shouldShowLogin: boolean;
  showLogin: (trigger: string) => void;
  closeLogin: () => void;
  loginState?: LoginState;
  logout: () => Promise<void>;
  updateUser: (user: LoggedUser) => Promise<void>;
  loadingUser?: boolean;
  tokenRefreshed: boolean;
  loadedUserFromCache?: boolean;
  getRedirectUri: () => string;
  anonymous?: AnonymousUser;
  visit?: Visit;
  isFirstVisit?: boolean;
  deleteAccount?: () => Promise<void>;
}

const AuthContext = React.createContext<AuthContextData>(null);
export default AuthContext;

const getQueryParams = () => {
  const urlSearchParams = new URLSearchParams(window.location.search);
  const params = Object.fromEntries(urlSearchParams.entries());

  return params;
};

export const REGISTRATION_PATH = '/register';

const logout = async (): Promise<void> => {
  await dispatchLogout();
  const params = getQueryParams();
  if (params.redirect_uri) {
    window.location.replace(params.redirect_uri);
  } else if (window.location.pathname === REGISTRATION_PATH) {
    window.location.replace(process.env.NEXT_PUBLIC_WEBAPP_URL);
  } else {
    window.location.reload();
  }
};

export type AuthContextProviderProps = {
  user: LoggedUser | AnonymousUser | undefined;
  children?: ReactNode;
} & Pick<
  AuthContextData,
  | 'getRedirectUri'
  | 'updateUser'
  | 'loadingUser'
  | 'tokenRefreshed'
  | 'loadedUserFromCache'
  | 'visit'
>;

export const AuthContextProvider = ({
  children,
  user,
  updateUser,
  loadingUser,
  tokenRefreshed,
  loadedUserFromCache,
  getRedirectUri,
  visit,
}: AuthContextProviderProps): ReactElement => {
  const [loginState, setLoginState] = useState<LoginState | null>(null);

  const authContext: AuthContextData = useMemo(
    () => ({
      user: user && 'providers' in user ? user : null,
      isFirstVisit: user?.isFirstVisit ?? false,
      trackingId: user?.id,
      shouldShowLogin: loginState !== null,
      showLogin: (trigger) => setLoginState({ trigger }),
      closeLogin: () => setLoginState(null),
      loginState,
      updateUser,
      logout,
      loadingUser,
      tokenRefreshed,
      loadedUserFromCache,
      getRedirectUri,
      anonymous: user,
      visit,
      deleteAccount,
    }),
    [user, loginState, loadingUser, tokenRefreshed, loadedUserFromCache, visit],
  );

  return (
    <AuthContext.Provider value={authContext}>{children}</AuthContext.Provider>
  );
};
