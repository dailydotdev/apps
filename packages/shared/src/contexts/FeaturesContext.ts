import React from 'react';
import { IFlags } from 'flagsmith';

export interface FeaturesData {
  flags: IFlags;
}

const FeaturesContext = React.createContext<FeaturesData>({ flags: {} });
export default FeaturesContext;
