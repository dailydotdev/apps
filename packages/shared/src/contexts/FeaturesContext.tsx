import React, { ReactElement, ReactNode, useMemo } from 'react';
import { IFlags } from 'flagsmith';
import {
  Features,
  getFeatureValue,
  isFeaturedEnabled,
} from '../lib/featureManagement';

export interface FeaturesData {
  flags: IFlags;
  showCommentPopover?: boolean;
  shouldShowMyFeed?: boolean;
  postEngagementNonClickable?: boolean;
  postModalByDefault?: boolean;
  postCardVersion?: string;
  additionalInteractionButtonFeature?: string;
}

const FeaturesContext = React.createContext<FeaturesData>({ flags: {} });
export default FeaturesContext;

export type FeaturesContextProviderProps = {
  children?: ReactNode;
  flags: IFlags | undefined;
};

export const FeaturesContextProvider = ({
  children,
  flags,
}: FeaturesContextProviderProps): ReactElement => {
  const features = useMemo(
    () => ({
      flags,
      showCommentPopover: isFeaturedEnabled(Features.ShowCommentPopover, flags),
      shouldShowMyFeed: isFeaturedEnabled(Features.MyFeedOn, flags) ?? true,
      postEngagementNonClickable: isFeaturedEnabled(
        Features.PostEngagementNonClickable,
        flags,
      ),
      postModalByDefault: isFeaturedEnabled(Features.PostModalByDefault, flags),
      postCardVersion: getFeatureValue(Features.PostCardVersion, flags),
      additionalInteractionButtonFeature: getFeatureValue(
        Features.AdditionalInteractionButton,
        flags,
      ),
    }),
    [flags],
  );

  return (
    <FeaturesContext.Provider value={features}>
      {children}
    </FeaturesContext.Provider>
  );
};
