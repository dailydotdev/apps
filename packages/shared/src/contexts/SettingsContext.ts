import React from 'react';
import { Spaciness } from '../graphql/settings';

export type SettingsContextData = {
  spaciness: Spaciness;
  themeMode: string;
  showOnlyUnreadPosts: boolean;
  openNewTab: boolean;
  insaneMode: boolean;
  setTheme: (theme) => Promise<void>;
  toggleShowOnlyUnreadPosts: () => Promise<void>;
  toggleOpenNewTab: () => Promise<void>;
  setSpaciness: (density: Spaciness) => Promise<void>;
  toggleInsaneMode: () => Promise<void>;
  loadedSettings: boolean;
};

const SettingsContext = React.createContext<SettingsContextData>(null);
export default SettingsContext;
