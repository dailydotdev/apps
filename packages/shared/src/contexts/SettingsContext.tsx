import React, {
  ReactElement,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
} from 'react';
import { useMutation } from 'react-query';
import request from 'graphql-request';
import {
  RemoteSettings,
  RemoteTheme,
  Spaciness,
  UPDATE_USER_SETTINGS_MUTATION,
} from '../graphql/settings';
import AuthContext from './AuthContext';
import { apiUrl } from '../lib/config';
import { storageWrapper } from '../lib/storageWrapper';

export enum ThemeMode {
  Light = 'light',
  Dark = 'dark',
  Auto = 'auto',
}

export type BookmarksSharing = {
  enabled: boolean;
  slug: string;
  rssUrl: string;
};

export type SettingsContextData = {
  spaciness: Spaciness;
  themeMode: ThemeMode;
  showOnlyUnreadPosts: boolean;
  openNewTab: boolean;
  insaneMode: boolean;
  showTopSites: boolean;
  sidebarExpanded: boolean;
  sortingEnabled: boolean;
  optOutWeeklyGoal: boolean;
  optOutCompanion: boolean;
  autoDismissNotifications: boolean;
  setTheme: (theme: ThemeMode) => Promise<void>;
  toggleShowOnlyUnreadPosts: () => Promise<void>;
  toggleOpenNewTab: () => Promise<void>;
  setSpaciness: (density: Spaciness) => Promise<void>;
  toggleInsaneMode: () => Promise<void>;
  toggleShowTopSites: () => Promise<void>;
  toggleSidebarExpanded: () => Promise<void>;
  toggleSortingEnabled: () => Promise<void>;
  toggleOptOutWeeklyGoal: () => Promise<void>;
  toggleOptOutCompanion: () => Promise<void>;
  toggleAutoDismissNotifications: () => Promise<void>;
  loadedSettings: boolean;
  customLinks?: string[];
  updateCustomLinks: (links: string[]) => Promise<unknown>;
};

const SettingsContext = React.createContext<SettingsContextData>(null);
export default SettingsContext;

const deprecatedLightModeStorageKey = 'showmethelight';

export const themeModes: Record<RemoteTheme, ThemeMode> = {
  bright: ThemeMode.Light,
  darcula: ThemeMode.Dark,
  auto: ThemeMode.Auto,
};

export const remoteThemes: Record<ThemeMode, RemoteTheme> = {
  [ThemeMode.Light]: 'bright',
  [ThemeMode.Dark]: 'darcula',
  [ThemeMode.Auto]: 'auto',
};

export function applyTheme(
  themeMode: ThemeMode,
  el: HTMLElement = document.documentElement,
): void {
  if (!el || el.classList.contains(themeMode)) {
    return;
  }

  if (themeMode === ThemeMode.Dark) {
    el.classList.remove(ThemeMode.Light);
    el.classList.remove(ThemeMode.Auto);
  } else if (themeMode === ThemeMode.Light) {
    el.classList.add(ThemeMode.Light);
    el.classList.remove(ThemeMode.Auto);
  } else {
    el.classList.remove(ThemeMode.Light);
    el.classList.add(ThemeMode.Auto);
  }
}

export type SettingsContextProviderProps = {
  children?: ReactNode;
  settings?: RemoteSettings;
  updateSettings?: (settings: RemoteSettings) => unknown;
  loadedSettings?: boolean;
};

const defaultSettings: RemoteSettings = {
  spaciness: 'eco',
  showOnlyUnreadPosts: false,
  openNewTab: true,
  insaneMode: false,
  showTopSites: true,
  sidebarExpanded: true,
  companionExpanded: false,
  sortingEnabled: false,
  optOutWeeklyGoal: false,
  optOutCompanion: false,
  autoDismissNotifications: true,
  theme: remoteThemes[ThemeMode.Dark],
};

export const SettingsContextProvider = ({
  children,
  settings = defaultSettings,
  updateSettings,
  loadedSettings,
}: SettingsContextProviderProps): ReactElement => {
  const { user } = useContext(AuthContext);
  const userId = user?.id;

  useEffect(() => {
    if (!loadedSettings) {
      return;
    }

    applyTheme(themeModes[settings.theme]);
  }, [settings.theme, loadedSettings]);

  const { mutateAsync: updateRemoteSettings } = useMutation<
    unknown,
    unknown,
    RemoteSettings
  >(
    (params) =>
      request(`${apiUrl}/graphql`, UPDATE_USER_SETTINGS_MUTATION, {
        data: params,
      }),
    {
      onError: (_, params) => {
        const rollback = Object.keys(params).reduce(
          (values, key) => ({ ...values, [key]: settings[key] }),
          {},
        );

        updateSettings({ ...settings, ...rollback });
      },
    },
  );

  useEffect(() => {
    const lightMode = storageWrapper.getItem(deprecatedLightModeStorageKey);
    if (lightMode === 'true') {
      applyTheme(ThemeMode.Light);
    }
  }, []);

  const updateRemoteSettingsFn = async (
    newSettings: RemoteSettings,
  ): Promise<void> => {
    if (userId) {
      await updateRemoteSettings({
        ...newSettings,
      });
    }
  };

  const setSettings = async (newSettings: RemoteSettings): Promise<void> => {
    updateSettings({ ...settings, ...newSettings });
    await updateRemoteSettingsFn(newSettings);
  };

  const contextData = useMemo<SettingsContextData>(
    () => ({
      ...settings,
      themeMode: themeModes[settings.theme],
      setTheme: (theme: ThemeMode) =>
        setSettings({ ...settings, theme: remoteThemes[theme] }),
      toggleShowOnlyUnreadPosts: () =>
        setSettings({
          ...settings,
          showOnlyUnreadPosts: !settings.showOnlyUnreadPosts,
        }),
      toggleOpenNewTab: () =>
        setSettings({ ...settings, openNewTab: !settings.openNewTab }),
      setSpaciness: (density) =>
        setSettings({ ...settings, spaciness: density }),
      toggleInsaneMode: () =>
        setSettings({ ...settings, insaneMode: !settings.insaneMode }),
      toggleShowTopSites: () =>
        setSettings({ ...settings, showTopSites: !settings.showTopSites }),
      toggleSidebarExpanded: () =>
        setSettings({
          ...settings,
          sidebarExpanded: !settings.sidebarExpanded,
        }),
      toggleSortingEnabled: () =>
        setSettings({ ...settings, sortingEnabled: !settings.sortingEnabled }),
      toggleOptOutWeeklyGoal: () =>
        setSettings({
          ...settings,
          optOutWeeklyGoal: !settings.optOutWeeklyGoal,
        }),
      toggleOptOutCompanion: () =>
        setSettings({
          ...settings,
          optOutCompanion: !settings.optOutCompanion,
        }),
      toggleAutoDismissNotifications: () =>
        setSettings({
          ...settings,
          autoDismissNotifications: !settings.autoDismissNotifications,
        }),
      loadedSettings,
      updateCustomLinks: (links: string[]) =>
        setSettings({ ...settings, customLinks: links }),
    }),
    [settings, loadedSettings, userId],
  );

  return (
    <SettingsContext.Provider value={contextData}>
      {children}
    </SettingsContext.Provider>
  );
};
