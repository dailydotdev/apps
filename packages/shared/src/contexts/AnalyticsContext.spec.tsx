import React, { ReactElement, ReactNode, useContext, useEffect } from 'react';
import nock from 'nock';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import AnalyticsContext, { AnalyticsContextProvider } from './AnalyticsContext';
import { AnalyticsContextData } from '../hooks/analytics/useAnalyticsContextData';
import FeaturesContext, { FeaturesData } from './FeaturesContext';
import SettingsContext, { SettingsContextData } from './SettingsContext';
import AuthContext, { AuthContextData } from './AuthContext';
import { AnonymousUser } from '../lib/user';
import { AnalyticsEvent } from '../hooks/analytics/useAnalyticsQueue';
import { Visit } from '../lib/boot';
import { waitForNock } from '../../__tests__/helpers/utilities';

let queryClient: QueryClient;
const getPage = jest.fn();

beforeEach(() => {
  jest.clearAllMocks();
  nock.cleanAll();
  queryClient = new QueryClient();
});

const features: FeaturesData = {
  flags: {
    login: {
      enabled: true,
      value: 'primary',
    },
    theme: {
      enabled: true,
      value: 'dark',
    },
  },
};

const settings: SettingsContextData = {
  spaciness: 'roomy',
  loadedSettings: true,
  setTheme: jest.fn(),
  themeMode: 'light',
  insaneMode: true,
  showTopSites: true,
  toggleInsaneMode: jest.fn(),
  openNewTab: true,
  setSpaciness: jest.fn(),
  toggleOpenNewTab: jest.fn(),
  toggleShowOnlyUnreadPosts: jest.fn(),
  toggleShowTopSites: jest.fn(),
  showOnlyUnreadPosts: false,
};

const AnalyticsContextTester = ({
  callback,
}: {
  callback: (contextData: AnalyticsContextData) => unknown;
}): ReactElement => {
  const contextData = useContext(AnalyticsContext);

  useEffect(() => {
    callback(contextData);
  }, []);

  return <></>;
};

const TestComponent = ({
  children,
  authContext,
}: {
  children: ReactNode;
  authContext: Pick<
    AuthContextData,
    'user' | 'anonymous' | 'tokenRefreshed' | 'visit'
  >;
}): ReactElement => (
  <QueryClientProvider client={queryClient}>
    <FeaturesContext.Provider value={features}>
      <AuthContext.Provider
        value={{
          shouldShowLogin: false,
          showLogin: jest.fn(),
          logout: jest.fn(),
          updateUser: jest.fn(),
          getRedirectUri: jest.fn(),
          closeLogin: jest.fn(),
          trackingId: authContext.user?.id || authContext.anonymous?.id,
          ...authContext,
        }}
      >
        <SettingsContext.Provider value={settings}>
          <AnalyticsContextProvider app="test" getPage={getPage} deviceId="123">
            {children}
          </AnalyticsContextProvider>
        </SettingsContext.Provider>
      </AuthContext.Provider>
    </FeaturesContext.Provider>
  </QueryClientProvider>
);

const baseAnonymous: AnonymousUser = {
  id: 'u',
  referrer: 'ido',
  firstVisit: new Date(Date.UTC(2021, 7, 28)).toISOString(),
};

const baseVisit: Visit = {
  visitId: 'v',
  sessionId: 's',
};

const mockEventsEndpoint = (takeSnapshot = true) => {
  nock('http://localhost:3000')
    .post('/e', (body: { events: AnalyticsEvent[] }) => {
      if (takeSnapshot) {
        // Reset time based properties
        expect(
          body.events.map((event) => {
            expect(event.event_id).toBeDefined();
            expect(event.event_timestamp).toBeDefined();
            expect(event.device_id).toBeDefined();
            return {
              ...event,
              event_id: '',
              event_timestamp: '',
              device_id: '',
            };
          }),
        ).toMatchSnapshot();
      }
      return true;
    })
    .reply(200, 'OK');
};

it('should batch events before sending', async () => {
  mockEventsEndpoint();
  const callback = ({ trackEvent }: AnalyticsContextData) => {
    trackEvent({ event_name: 'e1' });
    trackEvent({ event_name: 'e2' });
  };

  render(
    <TestComponent
      authContext={{
        anonymous: baseAnonymous,
        tokenRefreshed: true,
        visit: baseVisit,
      }}
    >
      <AnalyticsContextTester callback={callback} />
    </TestComponent>,
  );
  await waitForNock();
});

it('should add relevant properties when user is signed-in', async () => {
  mockEventsEndpoint();
  const callback = ({ trackEvent }: AnalyticsContextData) => {
    trackEvent({ event_name: 'e1' });
  };
  render(
    <TestComponent
      authContext={{
        anonymous: baseAnonymous,
        visit: baseVisit,
        user: {
          id: 'u',
          createdAt: new Date(Date.UTC(2021, 7, 29)).toISOString(),
          email: 'u@a.com',
          image: 'https://image.com',
          name: 'U',
          providers: ['github'],
        },
        tokenRefreshed: true,
      }}
    >
      <AnalyticsContextTester callback={callback} />
    </TestComponent>,
  );
  await waitForNock();
});

it('should send events in different batches', async () => {
  let done = false;
  const callback = async ({ trackEvent }: AnalyticsContextData) => {
    mockEventsEndpoint();
    trackEvent({ event_name: 'e1' });
    await new Promise((resolve) => setTimeout(resolve, 600));
    mockEventsEndpoint();
    trackEvent({ event_name: 'e2' });
    done = true;
  };

  render(
    <TestComponent
      authContext={{
        anonymous: baseAnonymous,
        tokenRefreshed: true,
        visit: baseVisit,
      }}
    >
      <AnalyticsContextTester callback={callback} />
    </TestComponent>,
  );
  await waitFor(() => expect(done).toBeTruthy());
  await waitForNock();
});

it('should send event with duration', async () => {
  nock('http://localhost:3000')
    .post('/e', (body: { events: AnalyticsEvent[] }) => {
      expect(body.events[0].event_duration).toBeTruthy();
      return true;
    })
    .reply(200, 'OK');

  const callback = async ({
    trackEventStart,
    trackEventEnd,
  }: AnalyticsContextData) => {
    trackEventStart('event', { event_name: 'e1' });
    await new Promise((resolve) => setTimeout(resolve, 10));
    trackEventEnd('event');
  };

  render(
    <TestComponent
      authContext={{
        anonymous: baseAnonymous,
        tokenRefreshed: true,
        visit: baseVisit,
      }}
    >
      <AnalyticsContextTester callback={callback} />
    </TestComponent>,
  );
  await waitForNock();
});

it('should send pending events when page becomes invisible', async () => {
  window.navigator.sendBeacon = jest.fn();

  const callback = async ({ trackEventStart }: AnalyticsContextData) => {
    trackEventStart('event', { event_name: 'e1' });
    await new Promise((resolve) => setTimeout(resolve, 10));
    fireEvent(
      window,
      new CustomEvent('statechange', {
        bubbles: true,
        detail: {
          oldState: 'active',
          newState: 'hidden',
        },
      }),
    );
  };

  render(
    <TestComponent
      authContext={{
        anonymous: baseAnonymous,
        tokenRefreshed: true,
        visit: baseVisit,
      }}
    >
      <AnalyticsContextTester callback={callback} />
    </TestComponent>,
  );
  await waitFor(() => expect(window.navigator.sendBeacon).toBeCalledTimes(1));
});

it('should send pending events when user information is fetched', async () => {
  let done = false;
  const callback = async ({ trackEvent }: AnalyticsContextData) => {
    trackEvent({ event_name: 'e1' });
    // Wait for debounce to finish
    await new Promise((resolve) => setTimeout(resolve, 100));
    done = true;
  };

  const { rerender } = render(
    <TestComponent
      authContext={{
        anonymous: baseAnonymous,
        tokenRefreshed: false,
        visit: baseVisit,
      }}
    >
      <AnalyticsContextTester callback={callback} />
    </TestComponent>,
  );
  await waitFor(() => expect(done).toBeTruthy());
  mockEventsEndpoint();
  rerender(
    <TestComponent
      authContext={{
        anonymous: {
          ...baseAnonymous,
          id: 'u2',
        },
        tokenRefreshed: true,
        visit: {
          sessionId: 's2',
          visitId: 'v2',
        },
      }}
    >
      <AnalyticsContextTester callback={callback} />
    </TestComponent>,
  );
  await waitForNock();
});
