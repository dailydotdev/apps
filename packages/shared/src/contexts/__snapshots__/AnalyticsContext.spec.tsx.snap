// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add relevant properties when user is signed-in 1`] = `
Array [
  Object {
    "app_platform": "test",
    "app_theme": "light",
    "device_id": "",
    "event_id": "",
    "event_name": "e1",
    "event_timestamp": "",
    "feature_flags": "{\\"login\\":{\\"enabled\\":true,\\"value\\":\\"primary\\"},\\"theme\\":{\\"enabled\\":true,\\"value\\":\\"dark\\"}}",
    "feed_density": "roomy",
    "feed_layout": "list",
    "page_referrer": "",
    "page_state": "passive",
    "screen_height": 0,
    "screen_width": 0,
    "session_id": "s",
    "user_first_visit": "2021-08-28T00:00:00.000Z",
    "user_id": "u",
    "user_referrer": "ido",
    "user_registration_date": "2021-08-29T00:00:00.000Z",
    "visit_id": "v",
    "window_height": 768,
    "window_width": 1024,
  },
]
`;

exports[`should batch events before sending 1`] = `
Array [
  Object {
    "app_platform": "test",
    "app_theme": "light",
    "device_id": "",
    "event_id": "",
    "event_name": "e1",
    "event_timestamp": "",
    "feature_flags": "{\\"login\\":{\\"enabled\\":true,\\"value\\":\\"primary\\"},\\"theme\\":{\\"enabled\\":true,\\"value\\":\\"dark\\"}}",
    "feed_density": "roomy",
    "feed_layout": "list",
    "page_referrer": "",
    "page_state": "passive",
    "screen_height": 0,
    "screen_width": 0,
    "session_id": "s",
    "user_first_visit": "2021-08-28T00:00:00.000Z",
    "user_id": "u",
    "user_referrer": "ido",
    "visit_id": "v",
    "window_height": 768,
    "window_width": 1024,
  },
  Object {
    "app_platform": "test",
    "app_theme": "light",
    "device_id": "",
    "event_id": "",
    "event_name": "e2",
    "event_timestamp": "",
    "feature_flags": "{\\"login\\":{\\"enabled\\":true,\\"value\\":\\"primary\\"},\\"theme\\":{\\"enabled\\":true,\\"value\\":\\"dark\\"}}",
    "feed_density": "roomy",
    "feed_layout": "list",
    "page_referrer": "",
    "page_state": "passive",
    "screen_height": 0,
    "screen_width": 0,
    "session_id": "s",
    "user_first_visit": "2021-08-28T00:00:00.000Z",
    "user_id": "u",
    "user_referrer": "ido",
    "visit_id": "v",
    "window_height": 768,
    "window_width": 1024,
  },
]
`;

exports[`should send events in different batches 1`] = `
Array [
  Object {
    "app_platform": "test",
    "app_theme": "light",
    "device_id": "",
    "event_id": "",
    "event_name": "e1",
    "event_timestamp": "",
    "feature_flags": "{\\"login\\":{\\"enabled\\":true,\\"value\\":\\"primary\\"},\\"theme\\":{\\"enabled\\":true,\\"value\\":\\"dark\\"}}",
    "feed_density": "roomy",
    "feed_layout": "list",
    "page_referrer": "",
    "page_state": "passive",
    "screen_height": 0,
    "screen_width": 0,
    "session_id": "s",
    "user_first_visit": "2021-08-28T00:00:00.000Z",
    "user_id": "u",
    "user_referrer": "ido",
    "visit_id": "v",
    "window_height": 768,
    "window_width": 1024,
  },
]
`;

exports[`should send events in different batches 2`] = `
Array [
  Object {
    "app_platform": "test",
    "app_theme": "light",
    "device_id": "",
    "event_id": "",
    "event_name": "e2",
    "event_timestamp": "",
    "feature_flags": "{\\"login\\":{\\"enabled\\":true,\\"value\\":\\"primary\\"},\\"theme\\":{\\"enabled\\":true,\\"value\\":\\"dark\\"}}",
    "feed_density": "roomy",
    "feed_layout": "list",
    "page_referrer": "",
    "page_state": "passive",
    "screen_height": 0,
    "screen_width": 0,
    "session_id": "s",
    "user_first_visit": "2021-08-28T00:00:00.000Z",
    "user_id": "u",
    "user_referrer": "ido",
    "visit_id": "v",
    "window_height": 768,
    "window_width": 1024,
  },
]
`;

exports[`should send pending events when user information is fetched 1`] = `
Array [
  Object {
    "app_platform": "test",
    "app_theme": "light",
    "device_id": "",
    "event_id": "",
    "event_name": "e1",
    "event_timestamp": "",
    "feature_flags": "{\\"login\\":{\\"enabled\\":true,\\"value\\":\\"primary\\"},\\"theme\\":{\\"enabled\\":true,\\"value\\":\\"dark\\"}}",
    "feed_density": "roomy",
    "feed_layout": "list",
    "page_referrer": "",
    "page_state": "passive",
    "screen_height": 0,
    "screen_width": 0,
    "session_id": "s2",
    "user_first_visit": "2021-08-28T00:00:00.000Z",
    "user_id": "u2",
    "user_referrer": "ido",
    "visit_id": "v2",
    "window_height": 768,
    "window_width": 1024,
  },
]
`;
